kotlin/samples/hello-terminal/terminalworkflow/src/main/java/com/squareup/sample/helloterminal/terminalworkflow/TerminalWorkflowRunner.kt

```diff
@@ -0,0 +1,156 @@
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.Channel.Factory.CONFLATED
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.selects.selectUnbiased
+import kotlinx.coroutines.withContext
+import kotlin.coroutines.EmptyCoroutineContext
+
+/**
+ * TODO write documentation.
+ */
+class TerminalWorkflowRunner(
+  private val hostFactory: WorkflowHost.Factory = WorkflowHost.Factory(EmptyCoroutineContext),
+  private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+) {
+
+  private val screen = DefaultTerminalFactory().createScreen()
+
+  @UseExperimental(ExperimentalCoroutinesApi::class, ObsoleteCoroutinesApi::class)
+  @Suppress("BlockingMethodInNonBlockingContext")
+  suspend fun run(workflow: TerminalWorkflow): ExitCode = coroutineScope {
+    val configs = Channel<TerminalConfig>(CONFLATED)
+    val wrapperWorkflow =
+      Workflow.stateless<TerminalConfig, ExitCode, TerminalRendering> { config ->
+        val originalRendering = renderChild(workflow, config) { emitOutput(it) }
+        return@stateless originalRendering.copy(
+            onKeyStroke = onEvent {
+              println("3: ${it.source}")
+              originalRendering.onKeyStroke(it)
+              noop()
+            }
+        )
+      }
+    val host = hostFactory.run(wrapperWorkflow, configs, context = coroutineContext)
+    val keyStrokes = screen.listenForKeyStrokesOn(this)
+    val resizes = screen.terminal.listenForResizesOn(this)
+
+    try {
+      withContext(ioDispatcher) {
+        screen.startScreen()
+        try {
+          runTerminalWorkflow(screen, configs, host, keyStrokes, resizes)
+        } finally {
+          screen.stopScreen()
+        }
+      }
+    } finally {
+      // Cancel all the coroutines we started so the coroutineScope block will actually exit if no
+      // exception was thrown.
+      host.updates.cancel()
+      keyStrokes.cancel()
+      resizes.cancel()
+    }
+  }
+}
+
+@Suppress("BlockingMethodInNonBlockingContext")
+private suspend fun runTerminalWorkflow(
+  screen: Screen,
+  configs: SendChannel<TerminalConfig>,
+  host: WorkflowHost<TerminalConfig, ExitCode, TerminalRendering>,
+  keyStrokes: ReceiveChannel<KeyStroke>,
+  resizes: ReceiveChannel<TerminalSize>
+): ExitCode {
+  var config = TerminalConfig(screen.terminalSize.toSize())
+  var lastRendering: TerminalRendering? = null
+
+  // Need to send an initial input for the workflow to start running.
+  configs.offer(config)
+
+  while (true) {
+    val update = selectUnbiased<Update<ExitCode, TerminalRendering>> {
+      keyStrokes.onReceive { key ->
+        println("2: $key")
+        lastRendering?.onKeyStroke?.invoke(key.toKeyStroke())
+
+        // Sending that event invalidated the lastRendering, so we don't want to re-iterate
```

Comment:
Good place to over-explain -- _Workflow renderings are able to accept one event / input change only, so sending that last event..._